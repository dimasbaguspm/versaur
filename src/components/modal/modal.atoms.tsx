import { forwardRef } from 'react'
import type {
  ModalHeaderProps,
  ModalFooterProps,
  ModalOverlayProps,
  ModalBodyProps,
} from './types'
import { cn } from '@/utils'

export const ModalHeader = forwardRef<HTMLDivElement, ModalHeaderProps>(
  ({ className, ...rest }, ref) => (
    <div
      ref={ref}
      className={cn('px-6 pt-6 pb-2 text-lg font-semibold', className)}
      {...rest}
    />
  )
)

export const ModalFooter = forwardRef<HTMLDivElement, ModalFooterProps>(
  ({ className, ...rest }, ref) => (
    <div
      ref={ref}
      className={cn('px-6 py-4  flex justify-end gap-2', className)}
      {...rest}
    />
  )
)

export const ModalBody = forwardRef<HTMLDivElement, ModalBodyProps>(
  ({ className, ...rest }, ref) => (
    <div ref={ref} className={cn('px-6 py-2', className)} {...rest} />
  )
)

/**
 * ModalOverlay atom: handles overlay rendering, click-to-close, and transitions
 */
export const ModalOverlay = forwardRef<HTMLDivElement, ModalOverlayProps>(
  ({ onOverlayClick, ...rest }, ref) => (
    <div
      ref={ref}
      role='presentation'
      tabIndex={-1}
      aria-modal='true'
      onClick={onOverlayClick}
      {...rest}
    />
  )
)
