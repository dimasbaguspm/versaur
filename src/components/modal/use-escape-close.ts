import { useEffect } from 'react'

/**
 * Calls handler when Escape key is pressed and active is true
 * @param active - Whether to listen for Escape key
 * @param handler - Function to call on Escape
 */
export function useEscapeClose(active: boolean, handler: () => void) {
  useEffect(() => {
    if (!active) return
    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Escape') handler()
    }
    document.addEventListener('keydown', handleKeyDown)
    return () => document.removeEventListener('keydown', handleKeyDown)
  }, [active, handler])
}
