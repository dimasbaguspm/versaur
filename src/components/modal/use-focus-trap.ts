import { useEffect } from 'react'

/**
 * Focuses the first focusable element inside the given ref when active is true
 * @param ref - The ref to the container element
 * @param active - Whether to activate the focus trap
 */
export function useFocusTrap(
  ref: React.RefObject<HTMLElement | null>,
  active: boolean
) {
  useEffect(() => {
    if (active && ref.current) {
      const focusable = ref.current.querySelectorAll<HTMLElement>(
        'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
      )
      if (focusable.length) focusable[0].focus()
    }
  }, [active, ref])
}
