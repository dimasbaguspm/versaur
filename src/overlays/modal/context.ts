import { createContext, useContext } from 'react'
import type { ModalRootProps } from './types'

/**
 * Context for the Modal compound component
 * Provides shared state and functions across all modal sub-components
 */
export interface ModalContextValue {
  isOpen: boolean
  onClose: () => void
  size?: ModalRootProps['size']
  placement?: ModalRootProps['placement']
}

export const ModalContext = createContext<ModalContextValue | null>(null)

/**
 * Hook to access modal context
 * Throws an error if used outside of a ModalRoot component
 */
export function useModalContext(): ModalContextValue {
  const context = useContext(ModalContext)
  if (!context) {
    throw new Error(
      'Modal components must be used within a ModalRoot component'
    )
  }
  return context
}
