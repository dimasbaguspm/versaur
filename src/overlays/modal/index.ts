export { Modal } from './modal'
export * from './types'
