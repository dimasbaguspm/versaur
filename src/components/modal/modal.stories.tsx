/**
 * Modal group stories
 *
 * - Demonstrates Modal compound usage
 * - Shows size and placement variants
 * - Follows accessibility best practices
 * - Usage matches README example
 */
import React from 'react'
import { Modal } from './modal'
import { Button } from '@/components/button'

export default {
  title: 'Components/Modal',
  component: Modal,
}

export function Default() {
  /**
   * Example usage of Modal compound component
   *
   * - Follows README usage
   * - Shows header, body, and footer
   */
  const [isOpen, setIsOpen] = React.useState(false)
  return (
    <>
      <Button onClick={() => setIsOpen(true)}>Open Modal</Button>
      <Modal isOpen={isOpen} onOpenChange={setIsOpen} aria-label='Dialog'>
        <Modal.Header>Title</Modal.Header>
        <Modal.Body>Body</Modal.Body>
        <Modal.Footer>
          <Button variant='ghost' onClick={() => setIsOpen(false)}>
            Cancel
          </Button>
          <Button variant='primary' onClick={() => setIsOpen(false)}>
            Confirm
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  )
}

function SizeModalExample({ size }: { size: 'sm' | 'md' | 'lg' }) {
  const [isOpen, setIsOpen] = React.useState(false)
  return (
    <div>
      <Button onClick={() => setIsOpen(true)}>Open {size}</Button>
      <Modal
        isOpen={isOpen}
        onOpenChange={setIsOpen}
        size={size}
        aria-label='Dialog'
      >
        <Modal.Header>{size.toUpperCase()} Modal</Modal.Header>
        <Modal.Body>Size: {size}</Modal.Body>
        <Modal.Footer>
          <Button variant='ghost' onClick={() => setIsOpen(false)}>
            Cancel
          </Button>
          <Button variant='primary' onClick={() => setIsOpen(false)}>
            Confirm
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  )
}

export const Sizes = () => (
  <div className='flex gap-4'>
    <SizeModalExample size='sm' />
    <SizeModalExample size='md' />
    <SizeModalExample size='lg' />
  </div>
)

function PlacementModalExample({ placement }: { placement: 'top' | 'center' }) {
  const [isOpen, setIsOpen] = React.useState(false)
  return (
    <div>
      <Button onClick={() => setIsOpen(true)}>Open {placement}</Button>
      <Modal
        isOpen={isOpen}
        onOpenChange={setIsOpen}
        placement={placement}
        aria-label='Dialog'
      >
        <Modal.Header>{placement.toUpperCase()} Modal</Modal.Header>
        <Modal.Body>Placement: {placement}</Modal.Body>
        <Modal.Footer>
          <Button variant='ghost' onClick={() => setIsOpen(false)}>
            Cancel
          </Button>
          <Button variant='primary' onClick={() => setIsOpen(false)}>
            Confirm
          </Button>
        </Modal.Footer>
      </Modal>
    </div>
  )
}

export const Placement = () => (
  <div className='flex gap-4'>
    <PlacementModalExample placement='top' />
    <PlacementModalExample placement='center' />
  </div>
)
