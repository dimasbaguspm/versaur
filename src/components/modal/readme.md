# Modal

A Versaur modal dialog component. Follows compound + context pattern, supports size and placement variants, and is fully accessible.

## Usage

```tsx
import { Modal } from '@/components/modal'
import { Button } from '@/components/button'

function Example() {
  const [isOpen, setIsOpen] = React.useState(false)
  return (
    <>
      <Button onClick={() => setIsOpen(true)}>Open Modal</Button>
      <Modal isOpen={isOpen} onOpenChange={setIsOpen} aria-label="Dialog">
        <Modal.Header>Title</Modal.Header>
        <Modal.Body className="px-6 py-4">Body</Modal.Body>
        <Modal.Footer>
          <Button variant="ghost" onClick={() => setIsOpen(false)}>
            Cancel
          </Button>
          <Button variant="primary" onClick={() => setIsOpen(false)}>
            Confirm
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  )
}
```
