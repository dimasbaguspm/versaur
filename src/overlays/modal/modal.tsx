import React from 'react'
import { cn } from '@/utils/cn'
import { ModalContext, type ModalContextValue } from './context'
import {
  ModalHeader,
  ModalFooter,
  ModalOverlay,
  ModalBody,
} from './modal.atoms'
import { modalContentVariants } from './helpers'
import { useEscapeClose } from '@/utils/use-escape-close'
import type { ModalRootProps } from './types'

/**
 * ModalRoot - A controlled modal overlay component
 * Provides shared state and context for modal compound parts
 */
export const ModalRoot: React.FC<ModalRootProps> = ({
  isOpen,
  onClose,
  size = 'md',
  placement = 'center',
  children,
  ...props
}) => {
  const contextValue = {
    isOpen,
    onClose,
    size,
    placement,
  } satisfies ModalContextValue

  const modalContentRef = useEscapeClose(isOpen, onClose)

  return (
    <ModalContext.Provider value={contextValue}>
      <div
        className={cn(
          'fixed z-50 inset-0 pointer-events-none',
          isOpen && 'pointer-events-auto'
        )}
      >
        <ModalOverlay />
        <div
          ref={modalContentRef}
          className={cn(modalContentVariants({ size, placement, isOpen }))}
          role='dialog'
          tabIndex={-1}
          aria-hidden={!isOpen}
          {...props}
        >
          {children}
        </div>
      </div>
    </ModalContext.Provider>
  )
}

/**
 * Modal - Compound component with sub-components attached
 * Provides a convenient API for using the modal with all its parts
 */
export const Modal = Object.assign(ModalRoot, {
  Header: ModalHeader,
  Body: ModalBody,
  Footer: ModalFooter,
})
