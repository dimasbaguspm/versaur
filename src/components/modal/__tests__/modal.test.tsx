import { describe, it, expect } from 'vitest'
import { render, fireEvent } from '@testing-library/react'
import * as stories from '../modal.stories'
import { composeStories } from '@storybook/react'

const { Default } = composeStories(stories)

/**
 * Modal unit tests
 *
 * - Uses composeStories from @storybook/react for consistency
 * - Asserts accessibility and html output
 * - Covers open/close via trigger, overlay, and ESC key
 */
describe('Modal', () => {
  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('opens and closes on trigger and overlay click', () => {
    const { getByText, queryByRole } = render(<Default />)
    fireEvent.click(getByText('Open Modal'))
    expect(queryByRole('dialog')).toBeInTheDocument()
    fireEvent.click(document.querySelector('[role="presentation"]')!)
    expect(queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('closes on ESC key', () => {
    const { getByText, queryByRole } = render(<Default />)
    fireEvent.click(getByText('Open Modal'))
    expect(queryByRole('dialog')).toBeInTheDocument()
    fireEvent.keyDown(document, { key: 'Escape' })
    expect(queryByRole('dialog')).not.toBeInTheDocument()
  })

  it('has accessible roles and labels', () => {
    const { getByText, getByRole } = render(<Default />)
    fireEvent.click(getByText('Open Modal'))
    const dialog = getByRole('dialog')
    expect(dialog).toHaveAttribute('aria-label', 'Dialog')
    expect(dialog).toBeVisible()
    expect(getByText('Title')).toBeInTheDocument()
    expect(getByText('Cancel')).toBeInTheDocument()
    expect(getByText('Confirm')).toBeInTheDocument()
  })
})
