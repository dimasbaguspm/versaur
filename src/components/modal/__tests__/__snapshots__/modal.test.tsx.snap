// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
    class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-coral text-white hover:bg-coral/90 focus-visible:ring-coral-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 px-4 text-sm min-w-[2.5rem]"
    type="button"
  >
    Open Modal
  </button>
  <div
    aria-modal="true"
    class="fixed inset-0 z-51 transition-opacity duration-300 bg-black/30 backdrop-blur-md flex items-center justify-center transition-opacity duration-200 ease-in-out items-center transition-opacity duration-200 ease-in-out opacity-0 pointer-events-none"
    placement="center"
    role="presentation"
    tabindex="-1"
  >
    <div
      aria-hidden="true"
      aria-label="Dialog"
      class="bg-white rounded-lg shadow-xl relative flex flex-col outline-none mx-4 transition-all duration-200 ease-in-out  w-[28rem] max-w-md transition-all duration-200 ease-in-out opacity-0 scale-95 pointer-events-none"
      role="dialog"
      tabindex="0"
    >
      <div
        class="px-6 pt-6 pb-2 text-lg font-semibold"
      >
        Title
      </div>
      <div
        class="px-6 py-2"
      >
        Body
      </div>
      <div
        class="px-6 py-4 flex justify-end gap-2"
      >
        <button
          aria-disabled="false"
          class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-transparent text-foreground hover:bg-slate-soft focus-visible:ring-slate-soft focus-visible:ring-offset-white h-10 px-4 text-sm min-w-[2.5rem]"
          type="button"
        >
          Cancel
        </button>
        <button
          aria-disabled="false"
          class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-coral text-white hover:bg-coral/90 focus-visible:ring-coral-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 px-4 text-sm min-w-[2.5rem]"
          type="button"
        >
          Confirm
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
